Examples/c++/Linked List/DLList.h

```diff
@@ -0,0 +1,401 @@
+
+		}
+
+		void Reverse()
+		{
+			NodePtr current = m_SentinelNode->Next;
+
+			while (current != m_SentinelNode)
+			{
+				std::swap(current->Next, current->Previous);
+				current = current->Previous;
+			}
+
+			std::swap(m_SentinelNode->Next, m_SentinelNode->Previous);
+			
+			m_SentinelNode->Previous->Next = m_SentinelNode;
+			m_SentinelNode->Next->Previous = m_SentinelNode;
+		}
+
+		void Reverse(Iterator from, Iterator to)
+		{
+			if (from == to)
+				return;
+
+			NodePtr beginNode = (NodePtr)from;
+			NodePtr endNode = ((NodePtr)to)->Previous;
+			NodePtr beginNodePrevious = beginNode->Previous;
+			NodePtr endNodeNext = (NodePtr)to;
+
+			endNode->Next = nullptr;
+
+			NodePtr current = beginNode;
+			NodePtr prev = nullptr;
+
+			while (current != nullptr)
+			{
+				std::swap(current->Next, current->Previous);
+				prev = current;
+				current = current->Previous;
+			}
+
+			beginNodePrevious->Next = prev;
+			beginNode->Next = endNodeNext;
+		}
+
+		void Clear()
+		{
+			if (m_Size == 0)
+				return;
+
+			NodePtr current = m_SentinelNode->Next;
+
+			while (current != m_SentinelNode)
+			{
+				NodePtr temp = current;
+				current = current->Next;
+				delete temp;
+			}
+
+			m_SentinelNode->Next = m_SentinelNode;
+			m_SentinelNode->Previous = m_SentinelNode;
+
+			m_Size = 0;
+		}
+
+		void Resize(int32_t size)
+		{
+			ASSERT(size >= 0, "Invalid size value!");
+
+			NodePtr current = m_SentinelNode->Previous;
+			int32_t difference = size - m_Size;
+
+			if (difference > 0)
+			{
+				for (size_t i = 0; i < difference; i++)
+					PushBack(T());
+			}
+			else if (difference < 0)
+			{
+				int32_t absoluteDifference = abs(difference);
+				for (size_t i = 0; i < absoluteDifference; i++)
+					PopBack();
+			}
+		}
+
+		size_t Size() const { return m_Size; }
+		bool Empty() const { return m_Size == 0; }
+
+		T& Head() const { return m_SentinelNode->Next->Data; }
+		T& Tail() const { return m_SentinelNode->Previous->Data; }
+
+		Iterator begin() { return Iterator(m_SentinelNode->Next, m_SentinelNode); }
+		Iterator end() { return Iterator(m_SentinelNode, m_SentinelNode); }
+	private:
+		void InsertBefore(NodePtr node, const T& data)
+		{
+			NodePtr newNode = new Node(data, node->Previous, node);
+			node->Previous->Next = newNode;
+			node->Previous = newNode;
+			m_Size++;
+		}
+
+		void InsertBefore(NodePtr node, T&& data)
+		{
+			NodePtr newNode = new Node(std::move(data), node->Previous, node);
+			node->Previous->Next = newNode;
+			node->Previous = newNode;
+			m_Size++;
+		}
+
+		void InsertAfter(NodePtr node, const T& data)
+		{
+			NodePtr newNode = new Node(data, node, node->Next);
+			node->Next->Previous = newNode;
+			node->Next = newNode;
+			m_Size++;
+		}
+
+		void InsertAfter(NodePtr node, T&& data)
+		{
+			NodePtr newNode = new Node(std::move(data), node, node->Next);
+			node->Next->Previous = newNode;
+			node->Next = newNode;
+			m_Size++;
+		}
+
+		template<typename... Args>
```

Comment:
For me these duplicates of the functions should not be done and some more comments should be added. Why actually this two functions even have a different behaviour. Could you explain a bit more?

Reply:
This is also important for object types. Similar to the STL function emplace and emplace_back this function constructs the object in-place given its parameters which is also faster than copying it.